### PR TITLE
PEX: Return capture group for matched patterns

### PR DESCRIPTION
**Pull request by reinkrul**
This can be used to extract string partials from a VC to return in Token Introspection. Use case: returning parsed CIBG X509 certificate SAN.otherName, which is a complex serial number in which the URA is included.

### CODE REVIEW

File: e2e-tests/oauth-flow/rfc021/docker-compose.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 services:
   nodeA-backend:
-    image: "${IMAGE_NODE_A:-nutsfoundation/nuts-node:master}"
+    image: "${IMAGE_NODE_A:-nutsfoundation/nuts-node:local}"
```
Comment (gerardsn) [inline e2e-tests/oauth-flow/rfc021/docker-compose.yml:3]:
```suggestion
    image: "${IMAGE_NODE_A:-nutsfoundation/nuts-node:master}"
```

File: e2e-tests/oauth-flow/rfc021/docker-compose.yml [hunk 2]
```diff
@@ -28,7 +28,7 @@ services:
       - "./node-A/html:/etc/nginx/html:ro"
       - "../scripts/oauth2.js:/etc/nginx/oauth2.js:ro"
   nodeB-backend:
-    image: "${IMAGE_NODE_B:-nutsfoundation/nuts-node:master}"
+    image: "${IMAGE_NODE_B:-nutsfoundation/nuts-node:local}"
```
Comment (gerardsn) [inline e2e-tests/oauth-flow/rfc021/docker-compose.yml:31]:
```suggestion
    image: "${IMAGE_NODE_B:-nutsfoundation/nuts-node:master}"
```

File: vcr/pe/presentation_definition.go [hunk 3]
```diff
@@ -473,70 +473,81 @@ func getValueAtPath(path string, vcAsInterface interface{}) (interface{}, error)
...
+		}
+		match, err := re.FindStringMatch(value.(string))
+		if err != nil {
+			return false, nil, err
 		}
-		return re.MatchString(value.(string))
+		if match == nil {
+			return false, nil, nil
+		}
+		// If there's a capture group, take the first one (not being the whole string, which is the first in the list)
```
Comment (gerardsn) [inline vcr/pe/presentation_definition.go:542]: I think this should fail if a filter contains more than 1 capture group. The PD is invalid?

File: docs/pages/deployment/pex.rst [hunk 4]
```diff
@@ -82,6 +82,19 @@ E.g., in the example above, a claim named ``fullName`` is added to the introspec
...
 (some fields are omitted for brevity):
 
+.. code-block:: json
+
+  {
+    "iss": "did:web:example.com",
+    "active": true,
+    "scope": "example_scope",
+    "fullName": "John Doe"
+  }
```
Comment (gerardsn) [inline docs/pages/deployment/pex.rst:92]: duplicate code blcok

File: docs/pages/deployment/pex.rst [hunk 5]
```diff
@@ -82,6 +82,19 @@ E.g., in the example above, a claim named ``fullName`` is added to the introspec
...
+
+If you want introspection to return part of a string, you can use the ``pattern`` regular expression filter in the field definition with a capture group.
+Token introspection will return the value of the first capture group in the regular expression, instead of the full match, e.g.;
+``{"role": "Admin level 4"}`` with the following pattern filter: ``"pattern": "Admin level ([0-9])"`` will return ``"role": "4"`` (given the field ID ``role``).
```
Comment (gerardsn) [inline docs/pages/deployment/pex.rst:96]: This is not enough information for me to understand what to do. Adding an example of a (partial) constraints field would help.

File: e2e-tests/oauth-flow/rfc021/node-A/presentationexchangemapping.json [hunk 6]
```diff
@@ -91,6 +91,17 @@
                 "filter": {
                   "type": "string"
                 }
+              },
+              {
+                "id": "employee_role_level",
```
Comment (gerardsn) [inline e2e-tests/oauth-flow/rfc021/node-A/presentationexchangemapping.json:96]: was the e2e test supposed to test the capture group?

File: vcr/pe/presentation_definition.go [hunk 7]
```diff
@@ -473,70 +473,81 @@ func getValueAtPath(path string, vcAsInterface interface{}) (interface{}, error)
...
 		for _, v := range typedValue {
-			match, err := matchFilter(filter, v)
+			match, _, err := matchFilter(filter, v)
 			if err != nil {
-				return false, err
+				return false, nil, err
 			}
 			if match {
-				return true, nil
+				return true, value, nil
```
Comment (gerardsn) [inline vcr/pe/presentation_definition.go:516]: should this return `v`?

File: vcr/pe/presentation_definition.go [hunk 8]
```diff
@@ -473,70 +473,81 @@ func getValueAtPath(path string, vcAsInterface interface{}) (interface{}, error)
 // Supported go value types: string, float64, int, bool and array.
 // 'null' values are not supported.
 // It returns an error on unsupported features or when the regex pattern fails.
-func matchFilter(filter Filter, value interface{}) (bool, error) {
+func matchFilter(filter Filter, value interface{}) (bool, interface{}, error) {
```
Comment (gerardsn) [inline vcr/pe/presentation_definition.go:476]: update godoc

**Review by gerardsn — CHANGES_REQUESTED** [review]

File: vcr/pe/presentation_definition.go (hunk 3 above)
Comment (reinkrul) [reply vcr/pe/presentation_definition.go:542]: I considered this, decided not to, but on second thought it's better to be more strict and make it more lenient if actually required.

File: vcr/pe/presentation_definition.go (hunk 7 above)
Comment (reinkrul) [reply vcr/pe/presentation_definition.go:516]: Before, it returned the whole slice if a slice entry matches. I wanted to retain that behavior (therefore, `value`, which is the whole slice).

File: e2e-tests/oauth-flow/rfc021/node-A/presentationexchangemapping.json (hunk 6 above)
Comment (reinkrul) [reply e2e-tests/oauth-flow/rfc021/node-A/presentationexchangemapping.json:96]: probably not the place to test it. Unit test is enough imo. Reverted it.

**Review by gerardsn — APPROVED** [review]

**Pull request merged**